File: 0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode buildTree(int left, int right, int[] nums) {
        if(left > right) return null;

        int mid = left + (right - left)/2;

        TreeNode node = new TreeNode(nums[mid]);

        if(left == right) return node;

        node.left = buildTree(left, mid-1, nums);
        node.right = buildTree(mid+1, right, nums);


        return node;
    }
    public TreeNode sortedArrayToBST(int[] nums) {
        return buildTree(0, nums.length-1, nums);
    }
}

File: 0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* iotToTree(vector<int> &nums,  int startPos, int endPos)
    {
        if(startPos>endPos)
            return NULL;
        int midPos = (startPos+endPos)/2;
        TreeNode* root= new TreeNode(nums[midPos]);
        root->left = iotToTree(nums, startPos, midPos-1);
        root->right = iotToTree(nums, midPos+1,endPos);
        return root;
    }
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        TreeNode* root = NULL;
        root = iotToTree(nums,0,nums.size()-1);
        return root;
    }
};